File: Leet_Weekly_Contest_279/Hoovely/Leet_2166.py
# leetcode 2166
# fix(idx) => idx에 있는 원소를 1로 변경
# unfix(idx) => idx에 있는 원소를 0으로 변경
# flip() => 0을 1로, 1을 0으로 변경
# all() => 모든 원소가 1이면 True 리턴, 아니면 False 리턴
# one() => 1이 하나라도 있으면 True 리턴, 아니면 False 리턴
# count() => 1갯수 count


class Bitset:
    bitset = []
    bit_num = 0

    def __init__(self, size: int):
        Bitset.bitset = [0]*size

    def fix(self, idx: int) -> None:
        Bitset.bitset[idx] = 1

    def unfix(self, idx: int) -> None:
        Bitset.bitset[idx] = 0

    def flip(self) -> None:
        for i in range(len(Bitset.bitset)):
            if Bitset.bitset[i] == 1:
                Bitset.bitset[i] = 0
            else:
                Bitset.bitset[i] = 1

    def all(self) -> bool:
        if 0 in Bitset.bitset:
            return False
        else:
            return True

    def one(self) -> bool:
        if 1 in Bitset.bitset:
            return True
        else:
            return False

    def count(self) -> int:
        return Bitset.bitset.count(1)

    def toString(self) -> str:
        bit_str = ""
        for i in Bitset.bitset:
            bit_str += str(i)
        return bit_str


# Your Bitset object will be instantiated and called as such:
# obj = Bitset(size)
# obj.fix(idx)
# obj.unfix(idx)
# obj.flip()
# param_4 = obj.all()
# param_5 = obj.one()
# param_6 = obj.count()
# param_7 = obj.toString()

File: Leet_Weekly_Contest_279/201824616/LEET_2166.cpp
class Bitset {
public:
    string s = "";
    string flip_s = "";
    int cnt = 0;
    bool f = 0;
    
    Bitset(int size) {
        for (int i = 0; i < size; i++) {
            s += "0";
            flip_s += "1";
        }
    }
    
    void fix(int idx) {
        if (!f && s[idx] == '0') {
            s.replace(idx, 1, "1");    
            flip_s.replace(idx, 1, "0");
            cnt++;
        }
        if (f && s[idx] == '1') {
            s.replace(idx, 1, "0");    
            flip_s.replace(idx, 1, "1");
            cnt++;
        }
    }
    
    void unfix(int idx) {
        if (!f && s[idx] == '1') {
            s.replace(idx, 1, "0");    
            flip_s.replace(idx, 1, "1");
            cnt--;
        }
        if (f && s[idx] == '0') {
            s.replace(idx, 1, "1");    
            flip_s.replace(idx, 1, "0");
            cnt--;
        }
    }
    
    void flip() {
        f = !f;
        cnt = s.length() - cnt;
    }
    
    bool all() {
        return (cnt == s.size());
    }
    
    bool one() {
        return (cnt > 0);
    }
    
    int count() {
        return cnt;
    }
    
    string toString() {
        if (!f) return s;
        return flip_s;
    }
};

/**
 * Your Bitset object will be instantiated and called as such:
 * Bitset* obj = new Bitset(size);
 * obj->fix(idx);
 * obj->unfix(idx);
 * obj->flip();
 * bool param_4 = obj->all();
 * bool param_5 = obj->one();
 * int param_6 = obj->count();
 * string param_7 = obj->toString();
 */
